### PR TITLE
Egmde

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Introduction to writing a Mir server

### CODE REVIEW

File: tutorials/desktop/example-mir-desktop/example-mir-desktop.md [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+published: 2018-05-15
+author: Alan Griffiths <alan@octopull.co.uk>
+
+---
+
+#  Egmde: Example Mir Desktop Environment (for Wayland)
+
+## Overview
+
+[Mir](https://mirserver.io) is a library designed to facilitate writing GUI shells for a range of platform from IoT and phones to desktops.
```
Comment (gerboland) [inline tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: platforms

File: tutorials/desktop/example-mir-desktop/example-mir-desktop.md [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+
+```bash
+sudo apt-add-repository ppa:mir-team/release
+sudo apt update
+```
+
+Install some development tools and the Mir libraries. It is also useful to install the `weston` package as the example makes use of `weston-terminal` as a Wayland based terminal application and the Qt toolkit's Wayland support : `qtwayland5`.
+
+```bash
+sudo apt install g++ cmake
```
Comment (gerboland) [inline tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: build-essential maybe?

Comment (gerboland) [reply tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: ah, might not match the symmetry with Fedora...

File: tutorials/desktop/example-mir-desktop/example-mir-desktop.md [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+
+```bash
+./egmde-desktop
+```
+
+You should see a blank screen with a `weston-terminal` session. From this you can run commands and, in particular, start graphical applications. Perhaps `qtcreator` to examine the code?
+
+![step-1](images/article-1.png)
+
+## Step 1: The code
```
Comment (gerboland) [inline tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: Step 2

File: tutorials/desktop/example-mir-desktop/example-mir-desktop.md [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+int main(int argc, char const* argv[])
+{
+    MirRunner runner{argc, argv};
+
+    return runner.run_with(
+        {
+            set_window_management_policy<egmde::WindowManagerPolicy>()
+        });
+}
+```
```
Comment (gerboland) [inline tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: Is the recap necessary?

File: tutorials/desktop/example-mir-desktop/example-mir-desktop.md [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+            CommandLineOption{
+                std::ref(wallpaper), 
+                "wallpaper", 
+                "Colour of wallpaper RGB", 
+                "0x92006a"},
+            StartupInternalClient{"wallpaper", std::ref(wallpaper)},
+            set_window_management_policy<egmde::WindowManagerPolicy>()
+        });
+}
+```
```
Comment (gerboland) [inline tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: Does markdown let you highlight the lines of code that you added?

File: tutorials/desktop/example-mir-desktop/example-mir-desktop.md [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+
+The `Keymap` utility handles setting the keymap, used like this it adds and uses a configuration option. (It can also be used to set a specific keymap on construction, or to change it dynamically while the shell is running.)
+
+The `CommandLineOption` utility does a number of things, adds a configuration option and calls it’s first argument with the configured value. To make this work nicely with an instance of egmde::Wallpaper the latter implements the function call operator to accept the wallpaper colour.
```
Comment (gerboland) [inline tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: its

File: tutorials/desktop/example-mir-desktop/example-mir-desktop.md (hunk 3 above)
Comment (gerboland) [reply tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: Your "step" numbering is confusing me, is this a sub-step of step 1?

File: tutorials/desktop/example-mir-desktop/example-mir-desktop.md [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+
+Alternatively, these options can be specified in a config file:
+
+```bash
+$ cat ~/.config/egmde.config 
+wallpaper=0xff0d0
+keymap=gb
+```
+
+## Step 2: The `main()` code
```
Comment (gerboland) [inline tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: another step 2?

File: tutorials/desktop/example-mir-desktop/example-mir-desktop.md [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+    MirBufferStream* buffer_stream = nullptr;
+    mir::client::Window window;
+
+    void create_window();
+    void handle_event(MirWindow* window, MirEvent const* ev);
+    static void handle_event(MirWindow* window, MirEvent const* event, void* context);
+};
+```
+
+In the current state of development Mir “internal clients” can only use the legacy Mir cliient API (not Wayland) but that is adequate for our current purpose.
```
Comment (gerboland) [inline tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: typo "client"

File: tutorials/desktop/example-mir-desktop/example-mir-desktop.md (hunk 5 above)
Comment (AlanGriffiths) [reply tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: Not that I can find

Comment (gerboland) [reply tutorials/desktop/example-mir-desktop/example-mir-desktop.md]: pity

**Pull request closed without merging**